### ISSUE TITLE
Topic-based retry support

### ISSUE THREAD

**Issue by henryjcee** [spring-projects/spring-kafka#920]
Retry support is (to my understanding) currently provided by Spring Retry. It would be nice to have built-in support for topic-based retries. See [here](https://eng.uber.com/reliable-reprocessing/) for more info on topic-based retries.

There is an [existing project](https://github.com/ZupIT/spring-kafka-retry) that has tried to do this but the approach could be simplified. Ideally, the `@KafkaListener` annotation could be configured parameterised to give full support with no other changes (assuming auto topic creation) e.g. 

`@KafkaListener(topics="test.topic", retries=5, backoff=Backoff.LINEAR, initialBackoff=60, dlq=true)`

Which would then autocreate the required topics (`test.topic.retry.60`, `test.topic.retry.120`, ..., `test.topic.dlq`) and create backoff-aware listeners for them (listeners that block until the retry period has expired, see the link above for more on this) that then call the annotated method with the backed-off message.

If there's support for this I can contribute.

**Comment by garyrussell** [comment]
I am not sure what changed, but it works fine with version 2.2.2 (boot 2.1.1); I'll look into why it doesn't work with 2.1.x tomorrow.

**Comment by henryjcee** [comment]
Sounds reasonable, I'll also take a look.

**Comment by nromanetti** [comment]
Just a remark, this pattern covers only the cases where ordering is not critical as you may loose the ordering that kafka guarantees. Assume that a message N fails, it will go to the retry topic... if  message N+1 succeeds before message N is successfully retried your order is lost.

**Comment by henryjcee** [comment]
Yes, it's essentially a way to achieve non-consumer group blocking retries at the expense of message ordering. It might seem a bit strange to tradeoff a key Kafka feature but I believe it's a widespread approach for people using Kafka for inter-service communication.

**Comment by lucas-piske** [comment]
Any progress on this issue? I think this feature might be really useful and I would like to contribute.

**Comment by henryjcee** [comment]
I started to have a look through the Spring retry source to see how this could be integrated but I quite quickly got out of my depth. Am still happy to contribute but I'll need to work alongside someone with Spring contribution experience.

**Comment by garyrussell** [comment]
The framework has had partial support for this since version 2.2.

See [Publishing Dead-letter Records](https://docs.spring.io/spring-kafka/docs/2.2.7.RELEASE/reference/html/#dead-letters).

You would have to manually set up the consumer(s) for the retry queue(s), however.

```
mainTopic -> main consumer -> failures sent to retryTopic (via dead letter publishing).
retryTopic -> retry consumer -> failures sent to final DLT
```

**Comment by artembilan** [comment]
See more progress in the https://github.com/spring-projects/spring-kafka/issues/1229

**Comment by mwesox** [comment]
any progress here?

**Comment by garyrussell** [comment]
Not so far; contributions are welcome.

**Comment by gururajnayak** [comment]
Any update on this issue. I would love to have this feature as suggested by the op without going through all the hoops.

**Comment by garyrussell** [comment]
The "existing project" in the original post is no longer there. 

If you have a working prototype of how this might work, we can take a look early next year.

**Comment by henryjcee** [comment]
@gururajnayak I am unfortunately no longer working with Spring and the code I started to write to implement this wasn't worth sharing I'm afraid. Sorry @garyrussell I feel like I've left you with more work than when I started!

**Comment by tomazfernandes** [comment]
@garyrussell, I've implemented something similar for my company recently, would it be alright if I provided the working prototype? I think I should have something sometime soon.

**Comment by tomazfernandes** [comment]
@garyrussell, I've implemented something similar recently, would it be alright if I provided the working prototype? I think I should have something sometime soon.

**Comment by garyrussell** [comment]
Yes, please do it in the form of a pull request so we can comment on GitHub. We won’t be back until Jan 4th.

**Comment by tomazfernandes** [comment]
Sure @garyrussell, I should have opened the PR by then, thanks a lot

**Comment by tomazfernandes** [comment]
@garyrussell done as requested, please feel free to ask me anything about the code / solution. Thanks!

**Comment by tomazfernandes** [comment]
Hi @garyrussell, I wonder if you had a chance to look into the PR I raised? If you have the time, do you think maybe we could talk about the high-level problem / solution and see if that approach makes sense? I'm ok with whatever you think is best, and any feedback will be appreciated. Also, no hurries if you're caught up with other tasks, just wanted to check in. Thanks!

**Comment by garyrussell** [comment]
@tomazfernandes Sorry; yes; too much stuff piled up over the long holiday (for us). I hope to look at this some time next week. We appreciate your efforts.

**Comment by tomazfernandes** [comment]
Sounds good @garyrussell, I too appreciate you having the time for looking into this. Let me know if there’s anything I can be of assistance on this matter. Thanks a lot.

**Comment by tomazfernandes** [comment]
Hi @garyrussell, just checking in again. DYT we might be able to talk about this issue anytime soon? If not, no worries, just trying not to have us forget about all this.

Also, if it’s too much code maybe we can just talk about the high level solution, and then perhaps I can change the prototype accordingly. It’d sure be a huge learning opportunity for me.

Thanks a lot for your great work.

**Comment by garyrussell** [comment]
Sorry - last week was tied up with getting new releases out for the Boot release; we also have to get out the 2.7 milestone tomorrow in time for the Boot 2.5 milestone, so I am cleaning up a final issue for that. I should be able to start looking at this after I build 2.7.0-M1 tomorrow.

Thanks for your patience.

**Comment by tomazfernandes** [comment]
No worries @garyrussell. I've resolved a conflict that had appeared, had to change some recent code for that, hope I didn't break anything =) Please let me know if I can be of any assistance. Thanks a lot.

**Comment by tomazfernandes** [comment]
For anyone in this thread that might not yet know, we've implemented this feature and it's been available since 2.7.0: https://docs.spring.io/spring-kafka/reference/html/#retry-topic

I'd like to thank everyone who contributed to this discussion since the solution was based in many aspects on what was presented here.

@garyrussell, while I do enjoy nostalgically remembering this issue, maybe it's time to close it? As you prefer, of course.

Thanks

**Comment by garyrussell** [comment]
Oops!
